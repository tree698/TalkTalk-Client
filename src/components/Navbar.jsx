import React from 'react';
import { useNavigate, Link } from 'react-router-dom';
import { AiOutlineHome } from 'react-icons/ai';
import { AiOutlineCloudUpload } from 'react-icons/ai';
import { AiOutlineLogout } from 'react-icons/ai';
import { useApiContext } from '../context/ApiContext';
import LoggedInUser from './LoggedInUser';

const STYLE_FOR_ICON_HOVER =
  'hover:rotate-17 hover:text-brand transition-all delay-150 duration-300 ease-in-out';
const STYLE_FOR_LOGGEDINUSER = 'flex items-center shrink-0';
const STYLE_FOR_AVATAR =
  'w-5 h-5 md:w-6 md:h-6 lg:w-7 lg:h-7 mr-0 lg:mr-1 rounded-full';
const STYLE_FOR_USERNAME = 'text-xs md:text-sm hidden lg:block';

export default function Navbar() {
  const { user, logOut } = useApiContext();
  const navigate = useNavigate();

  const handleClick = () => {
    if (window.confirm('Do you want to log out?')) {
      logOut().then(() => navigate('/'));
    }
  };

  return (
    <header
      style={{ boxShadow: '0px 1px 9px 3px rgba(124, 121, 121, 0.2)' }}
      className="w-full py-2 px-5 z-10 border-b flex justify-between border-superLightGray"
    >
      <div className="hover:brightness-110 transition-all delay-150 duration-300 ease-in-out">
        <Link to="/home" className="">
          <img
            src="images/logo.png"
            alt="logo"
            className="w-32 md:w-36 lg:w-40"
          />
        </Link>
      </div>
      <nav className="text-lg md:text-xl lg:text-2xl gap-1 md:gap-2 lg:gap-3 flex items-center">
        <Link to="/home">
          <AiOutlineHome className={STYLE_FOR_ICON_HOVER} />
        </Link>
        <Link to="/upload">
          <AiOutlineCloudUpload className={STYLE_FOR_ICON_HOVER} />
        </Link>
        <LoggedInUser
          photo={user.photo}
          username={user.username}
          styleForAvatar={STYLE_FOR_AVATAR}
          styleForUsername={STYLE_FOR_USERNAME}
          styleForLoggedInUser={STYLE_FOR_LOGGEDINUSER}
        />
        <button onClick={handleClick}>
          <AiOutlineLogout className={STYLE_FOR_ICON_HOVER} />
        </button>
      </nav>
    </header>
  );
}
