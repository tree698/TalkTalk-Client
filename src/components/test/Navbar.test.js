import renderer from 'react-test-renderer';
import { withApiContext, withRouter } from '../../test/utils';
import Navbar from '../Navbar';
import { Route } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

describe('Navbar', () => {
  let fakeContext;
  beforeEach(() => {
    fakeContext = {
      user: {
        photo: 'test',
        username: 'test',
      },
      logout: jest.fn(),
    };
  });
  afterEach(() => fakeContext.logout.mockReset());

  it('renders correctly', () => {
    const component = renderer.create(
      withApiContext(
        withRouter(<Route path="/" element={<Navbar />} />),
        fakeContext
      )
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('display confirm message and navigate to home when clicking logout button', () => {
    const mockAlert = jest
      .spyOn(window, 'confirm')
      .mockImplementation(() => {});

    render(
      withApiContext(
        withRouter(
          <>
            <Route path="/navbar" element={<Navbar />} />
            <Route path="/" element={<p>This is Home Page</p>} />
          </>,
          '/navbar'
        ),
        fakeContext
      )
    );
    const btn = screen.getByRole('button');

    userEvent.click(btn);

    expect(mockAlert).toHaveBeenCalled();
    // expect(screen.getByText('This is Home Page')).toBeInTheDocument();
  });
});
