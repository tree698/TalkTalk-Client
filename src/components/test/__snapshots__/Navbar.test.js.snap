// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders correctly 1`] = `
<header
  className="w-full py-2 px-5 z-10 border-b flex justify-between border-superLightGray"
  style={
    Object {
      "boxShadow": "0px 1px 9px 3px rgba(124, 121, 121, 0.2)",
    }
  }
>
  <div
    className="hover:brightness-110 transition-all delay-150 duration-300 ease-in-out"
  >
    <a
      href="/home"
      onClick={[Function]}
    >
      <img
        alt="logo"
        className="w-32 md:w-36 lg:w-40"
        src="images/logo.png"
      />
    </a>
  </div>
  <nav
    className="text-lg md:text-xl lg:text-2xl gap-1 md:gap-2 lg:gap-3 flex items-center"
  >
    <a
      href="/home"
      onClick={[Function]}
    >
      <svg
        className="hover:rotate-17 hover:text-brand transition-all delay-150 duration-300 ease-in-out"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 1024 1024"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M946.5 505L560.1 118.8l-25.9-25.9a31.5 31.5 0 0 0-44.4 0L77.5 505a63.9 63.9 0 0 0-18.8 46c.4 35.2 29.7 63.3 64.9 63.3h42.5V940h691.8V614.3h43.4c17.1 0 33.2-6.7 45.3-18.8a63.6 63.6 0 0 0 18.7-45.3c0-17-6.7-33.1-18.8-45.2zM568 868H456V664h112v204zm217.9-325.7V868H632V640c0-22.1-17.9-40-40-40H432c-22.1 0-40 17.9-40 40v228H238.1V542.3h-96l370-369.7 23.1 23.1L882 542.3h-96.1z"
        />
      </svg>
    </a>
    <a
      href="/upload"
      onClick={[Function]}
    >
      <svg
        className="hover:rotate-17 hover:text-brand transition-all delay-150 duration-300 ease-in-out"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 1024 1024"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M518.3 459a8 8 0 0 0-12.6 0l-112 141.7a7.98 7.98 0 0 0 6.3 12.9h73.9V856c0 4.4 3.6 8 8 8h60c4.4 0 8-3.6 8-8V613.7H624c6.7 0 10.4-7.7 6.3-12.9L518.3 459z"
        />
        <path
          d="M811.4 366.7C765.6 245.9 648.9 160 512.2 160S258.8 245.8 213 366.6C127.3 389.1 64 467.2 64 560c0 110.5 89.5 200 199.9 200H304c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8h-40.1c-33.7 0-65.4-13.4-89-37.7-23.5-24.2-36-56.8-34.9-90.6.9-26.4 9.9-51.2 26.2-72.1 16.7-21.3 40.1-36.8 66.1-43.7l37.9-9.9 13.9-36.6c8.6-22.8 20.6-44.1 35.7-63.4a245.6 245.6 0 0 1 52.4-49.9c41.1-28.9 89.5-44.2 140-44.2s98.9 15.3 140 44.2c19.9 14 37.5 30.8 52.4 49.9 15.1 19.3 27.1 40.7 35.7 63.4l13.8 36.5 37.8 10C846.1 454.5 884 503.8 884 560c0 33.1-12.9 64.3-36.3 87.7a123.07 123.07 0 0 1-87.6 36.3H720c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h40.1C870.5 760 960 670.5 960 560c0-92.7-63.1-170.7-148.6-193.3z"
        />
      </svg>
    </a>
    <div
      className="flex items-center shrink-0"
    >
      <div>
        <img
          alt="test"
          className="w-5 h-5 md:w-6 md:h-6 lg:w-7 lg:h-7 mr-0 lg:mr-1 rounded-full"
          src="test"
        />
      </div>
      <span
        className="text-xs md:text-sm hidden lg:block"
      >
        test
      </span>
    </div>
    <button
      onClick={[Function]}
    >
      <svg
        className="hover:rotate-17 hover:text-brand transition-all delay-150 duration-300 ease-in-out"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 1024 1024"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M868 732h-70.3c-4.8 0-9.3 2.1-12.3 5.8-7 8.5-14.5 16.7-22.4 24.5a353.84 353.84 0 0 1-112.7 75.9A352.8 352.8 0 0 1 512.4 866c-47.9 0-94.3-9.4-137.9-27.8a353.84 353.84 0 0 1-112.7-75.9 353.28 353.28 0 0 1-76-112.5C167.3 606.2 158 559.9 158 512s9.4-94.2 27.8-137.8c17.8-42.1 43.4-80 76-112.5s70.5-58.1 112.7-75.9c43.6-18.4 90-27.8 137.9-27.8 47.9 0 94.3 9.3 137.9 27.8 42.2 17.8 80.1 43.4 112.7 75.9 7.9 7.9 15.3 16.1 22.4 24.5 3 3.7 7.6 5.8 12.3 5.8H868c6.3 0 10.2-7 6.7-12.3C798 160.5 663.8 81.6 511.3 82 271.7 82.6 79.6 277.1 82 516.4 84.4 751.9 276.2 942 512.4 942c152.1 0 285.7-78.8 362.3-197.7 3.4-5.3-.4-12.3-6.7-12.3zm88.9-226.3L815 393.7c-5.3-4.2-13-.4-13 6.3v76H488c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h314v76c0 6.7 7.8 10.5 13 6.3l141.9-112a8 8 0 0 0 0-12.6z"
        />
      </svg>
    </button>
  </nav>
</header>
`;
